server.py

```diff
@@ -174,5 +182,68 @@ async def worker_setup(request):
     return json({"status": 200, "name": name, "token": new_token})
 
 
+# admin pages
+
+
+app.static("/static", "web/static")
+
+
+@app.get("/admin")
+async def admin_root(request):
+    if request.ctx.identity:
+        _, permission_level = await request.ctx.ensure_user()
+        return jinja2("admin/index.html", owner=(permission_level >= PermissionLevel.ADMIN))
+    else:
+        return jinja2("admin/login.html")
+
+@app.get("/admin/auth/<token>")
+async def admin_auth_token(request, token):
+    if request.ctx.identity:
+        return redirect("/admin")
+    identity = await database.check_webtoken(token)
+    session = await database.create_websession(identity)
+    response = redirect("/admin")
+    response.cookies["session"] = session
+    response.cookies["session"]["max-age"] = 24 * 60 * 60
+    response.cookies["session"]["secure"] = not DEBUG
+    response.cookies["session"]["httponly"] = True
+    response.cookies["session"]["path"] = "/admin"
+    return response
+
+# middlewares & utils
+
+def jinja2(template_name, **ctx):
+    async def render_stream(response):
+        stream_data = env.get_template(template_name).generate_async(**ctx)
+        async for chunk in stream_data:
+            await response.write(chunk)
+    return stream(render_stream, content_type="text/html")
+
+@app.middleware('response')
+async def prevent_xss(request, response):
+    response.headers["x-xss-protection"] = "1; mode=block"
+
+
+@app.middleware('request')
+async def admin_auth_middleware(request):
+    handler, args, kwargs, uri, name = app.router.get(request)
+    if name.startswith("admin_"):
+        print(request.cookies.get("session", None))
+        request.ctx.identity = await database.check_websession(request.cookies.get("session", None))
+        print(request.ctx.identity)
```

Comment:
Do we want this prints in production (change to logging.info then) or is it dev thing only.

Reply:
Dev, forgot to remove